subprojects/core/src/main/java/org/gradle/internal/service/scopes/TaskExecutionServices.java

```diff
@@ -114,21 +119,22 @@ TaskExecuter createTaskExecuter(TaskArtifactStateRepository repository,
                                     TaskExecutionGraphInternal taskExecutionGraph,
                                     BuildInvocationScopeId buildInvocationScopeId,
                                     BuildCancellationToken buildCancellationToken,
-                                    TaskExecutionListener taskExecutionListener
+                                    TaskExecutionListener taskExecutionListener,
+                                    TimeoutHandler timeoutHandler
     ) {
 
         boolean buildCacheEnabled = buildCacheController.isEnabled();
         boolean scanPluginApplied = buildScanPlugin.isBuildScanPluginApplied();
         TaskOutputChangesListener taskOutputChangesListener = listenerManager.getBroadcaster(TaskOutputChangesListener.class);
 
         TaskExecuter executer = new ExecuteActionsTaskExecuter(
-            taskOutputChangesListener,
-            listenerManager.getBroadcaster(TaskActionListener.class),
             buildOperationExecutor,
             asyncWorkTracker,
-            buildInvocationScopeId,
             buildCancellationToken
         );
+        executer = new ActionEventFiringTaskExecuter(executer, taskOutputChangesListener, listenerManager.getBroadcaster(TaskActionListener.class));
+        executer = new TimeoutTaskExecuter(executer, timeoutHandler);
+        executer = new SnapshotAfterExecutionTaskExecuter(executer, buildInvocationScopeId);
```

Comment:
As much as I hate nesting these decorators, and would like to use a different coding construct, I like that we have split this logic now.

Reply:
Why do you not like this idiom and what would you replace it with? 

I actually like it because it is similar to function composition. It could be a bit purer instead of mutating the passed arguments, but apart from that it is pretty readable. I know exactly what happens first by following the list here, without having to read the full implementation.

Reply:
I don't like it because it makes it hard to work with the data exposed to each executor. Instead of passing mutable contexts and states down the line, we should have method calls that take exactly the immutable arguments that are needed for the specific "executer" only.

I would keep each "executer's" `execute()` implementation roughly the same, but supply them with different argument types, and then tie them together in a `TaskExecutionManager` or something in a single method. This would allow the proper initialization and limited exposure of immutable data for each step.

Using standard conditionals to direct the flow would make it easier to follow which "executer" could stop/short-circuit execution and when, too.

Reply:
I'd very much like to keep a straight line of flow instead of a bunch of conditionals. The more functional the better.

+1 for passing along immutable data

Reply:
I'm not suggesting we should change the line of flow, it's just that the conditionals would not be part of each implementation, but the overarching "manager." Anyway, I'll spike something at some point and then we can discuss it further.